ProjFS.Mac/PrjFSKextLogDaemon/PrjFSKextLogDaemon.cpp

```diff
@@ -155,6 +165,123 @@ static os_log_type_t KextLogLevelAsOSLogType(KextLog_Level level)
+{
+    DIR *dr = opendir(PanicLogDirectory);
+    if (dr == NULL)
+    {
+        ostringstream errorMessage;
+        errorMessage << "Unable to open " << PanicLogDirectory << ", errno=" << errno << ", errorstr=" << strerror(errno);
+        LogDaemonError(errorMessage.str());
+    }
+
+    // Look up the last run time
+    long lastRunTimestamp = 0;
+    if (DoesPathExist(PanicLogTimestampFile))
+    {
+        lastRunTimestamp = GetFileModifiedTime(PanicLogTimestampFile);
+    }
+
+    struct dirent *dirEntry;
+    while ((dirEntry = readdir(dr)) != NULL)
+    {
+        // Only check files
+        if(dirEntry->d_type == DT_REG)
+        {
+           // Check for .panic extension
+           const char *ext = strrchr(dirEntry->d_name,'.');
+           if(ext && strcmp(ext, ".panic") == 0)
+           {
+               // Get the full name of the panic file
+               string fullFileName = PanicLogDirectory;
+               fullFileName += "/";
+               fullFileName += dirEntry->d_name;
+
+               // Verify it's after the last timestamp
+               long panicLogTime = GetFileModifiedTime(fullFileName.c_str());
+               if (panicLogTime == 0)
+               {
+                   // Record error if we can't determine timestamp
+                   ostringstream errorMessage;
+                   errorMessage << "Unable to run stat on panic at " << fullFileName.c_str() << "\n";
+                   const char *errorMessageString = errorMessage.str().c_str();
+                   LogKextMessage(OS_LOG_TYPE_DEFAULT, 0, errorMessageString, static_cast<int>(strlen(errorMessageString)));
+                   continue;
+               }
+
+               if (panicLogTime >= lastRunTimestamp)
+               {
+                  // Verify 'org.vfsforgit.PrjFSKext(' occurs in the log
+                  FILE *fp = fopen(fullFileName.c_str(), "r");
+                  if (fp)
+                  {
+                      char* line = NULL;
+                      size_t len = 0;
+                      ssize_t read;
+                      while ((read = getline(&line, &len, fp)) != -1)
+                      {
+                          if (strstr(line, "org.vfsforgit.PrjFSKext(") != NULL)
+                          {
+                              ostringstream errorMessage;
+                              errorMessage << "Found panic at " << fullFileName.c_str() << "\n";
+                              LogDaemonError(errorMessage.str());
+                          }
+                      }
+                      
+                      fclose(fp);
+                   
+                      if (line)
+                      {
+                          free(line);
+                      }
+                  }
+              }
+           }
+        }
+    }
+    closedir(dr);
+
+    // Update timestamp file
+    if (!DoesPathExist(PanicLogTimestampDirectory))
+    {
+        mkdir(PanicLogTimestampDirectory, 0777);
```

Comment:
Any reason to make this world-writable?

Reply:
By default all of our C# directories are 0777 (Directory.CreateDirectory default perms).  It's also the same as the Placeholder dir.  Considering what the directory, I don't think we need to limit access.